### DISCUSSION TITLE
Exception handling in ClassLoader::loadClass

### DISCUSSION THREAD

**Discussion by erickcomp** [composer/composer#11443] [Q&A / Support] [answered]
I'm in a project that's developing a package for other team of developers ("Users"). We use a structure that loads fallback classes(provided by our package) and lets the User override those classes providing their own definition by placing them in a PSR-4 compliant directory under the "app" directory. Everything works as expected, except by one edge case:

Note: Users are using an optimized autoloader (not authoritative, though);

1 - User creates a class that overrides package one;
2 - User installs a package and the class created at step one goes to classMap;
3 - User deletes the class created at step one;
4 - Composer autoloader tries to load class using the file path from classmap and boom! File does not exist.

I have found a possible solution, but that would require putting a try/catch block inside ClassLoader::loadClass when calling the $includeClass closure, as follows:

```php

// CURRENT:

/**
 * Loads the given class or interface.
 *
 * @param  string    $class The name of the class
 * @return true|null True if loaded, null otherwise
 */
public function loadClass($class)
{
    if ($file = $this->findFile($class)) {
        $includeFile = self::$includeFile;
        $includeFile($file);

        return true;
    }

    return null;
}

// MODIFIED:

/**
 * Loads the given class or interface.
 *
 * @param  string    $class The name of the class
 * @return true|null True if loaded, null otherwise
 */
public function loadClass($class)
{
    if ($file = $this->findFile($class)) {
        $includeFile = self::$includeFile;

        try {
            $includeFile($file);
        } catch (\Throwable $t) {
            return false;
        }

        return true;
    }

    return null;
}
```

If the method returns ```php false ``` at this point, it could let the developer register a fallback autoloader or let the exception raise. What do you guys think? I know that this try catch would only work if there's an error handler that changes errors into throwables.

Thanks for reading

**Comment by Seldaek** [comment]
This is a known "issue" or rather compromise of generating an optimized classloader. If you do that then you should not be adding/removing classes as they will be either not found or cause the issue you are mentioning.

IMO it works as advertised, and you should in development rather not use an optimized classloader.

**Reply by erickcomp** [reply]
I agree, but the fact is that those "Users" I have mentioned are using my package in a Laravel app, that uses optimized autoloader by default. For now, I'm gonna wrap composer's autoloader to make this exception stuff work.

**Reply by Seldaek** [reply, marked as answer]
Ok, well up to you. That sounds bad to me and I wish Laravel didn't ship with such defaults but I can see it's not really under your control.

**Reply by erickcomp** [reply]
Thanks!
